src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,862 @@
+ * \param[in]   com_map         This is the collision map used during the
+ *                              placement process for bookkeeping and
+ *                              determining which domains have been used.
+ * \param[in]   group_cnt       This number of redundancy groups for this
+ *                              objects layout.
+ * \param[in]   group_size      The size of each redundancy group in the layout.
+ *                              This will be multiplied with group_cnt to get
+ *                              the total number of targets required for
+ *                              placement.
+ * \param[in]   tree_depth      The size the stack will need to be when calling
+ *                              the iterative get_target function.
+ * \param[in]   dom_map_size    The number of non-leaf nodes in the pool map.
+ *                              This value will be used to initialize the array
+ *                              that holds the statuses of whether the node has
+ *                              been used or not.
+ * \param[in]   cnt_map_size    The number of child domains used for a given
+ *                              non-leaf node in the pool map. When this equals
+ *                              the total number of child nodes we reset the
+ *                              status showing the nodes have been used.
+ * \param[out]  remap_list      The list of nodes that needs to be rebuilt
+ *                              during the rebuild process. This is only
+ *                              returned when this function is called from the
+ *                              rebuild function and will otherwise be NULL.
+ *
+ * \return                      Returns the number of targets that need to be
+ *                              rebuilt. This number will be ignored when called
+ *                              from functions other than find rebuild.
+ */
+static int
+get_target_layout(struct pool_domain *root, struct pl_obj_layout *layout,
+		  struct coll_map *co_map, uint16_t group_cnt,
+		  uint16_t group_size, daos_obj_id_t oid, uint8_t tree_depth,
+		  uint32_t dom_map_size, uint32_t cnt_map_size,
+		  struct remap_node *remap_list)
+{
+	struct pool_target *target;
+	struct pool_domain *next_dom[tree_depth];
+	struct coll_map *next_co_map[tree_depth];
+	uint8_t dom_used[dom_map_size];
+	uint32_t dom_cnt[cnt_map_size];
+	uint32_t target_id;
+	uint32_t rebuild_doms[group_size * group_cnt];
+	uint8_t rebuild_shard_num[group_size * group_cnt];
+	uint32_t used_targets[layout->ol_nr * 2];
+	uint8_t rebuild_num = 0;
+	int i;
+	int j;
+	int k;
+
+	memset(dom_used, 0, sizeof(*dom_used) * dom_map_size);
+	memset(dom_cnt, 0, sizeof(*dom_cnt) *  cnt_map_size);
+	memset(used_targets, 0, (sizeof(*used_targets) * layout->ol_nr) * 2);
+
+	for (i = 0, k = 0; i < group_cnt; i++) {
+		for (j = 0; j < group_size; j++, k++) {
+			uint32_t tgt_id;
+			uint32_t ndom_id;
+
+			get_target(root, &target, co_map, crc(oid.lo, k),
```

Comment:
Now the 'VOS' is the final leaf target in pool map, so we need to make sure that shard within same redundancy group not located in same node (which is domain above VOS target in pool map), I don't see this is enforced in current code.

And should we consider the feature of specifying fault tolerance level? For instance, default replicas are located on different nodes, if a higher FT level (rack for instance) is specified (in object class?), replicas needs be located on different racks. I don't think we need to implement this in this patch, just please keep this in mind when coding. (for easier extension)

Reply:
Currently each layout location is a separate target.
This is enforced during target selection by only allowing the selection of domains that have not previously been selected.
 
Do we **only** need to confirm that they are on different nodes?
If so, how should we select targets for each shard?